### ISSUE TITLE
Exponentially slow compile time with repeated logsumexp when gradient is enabled

### ISSUE THREAD

**Issue by hkchengrex** [pytorch/pytorch#146663]
### 🐛 Describe the bug

I am trying to compile a simplified Sinkhorn algorithm. The core function is

```python
@torch.compile
def sinkhorn(Mr, u, v):
    for _ in range(num_iters):
        v = torch.logsumexp(Mr + u.unsqueeze(1), 0)
        u = torch.logsumexp(Mr + v.unsqueeze(0), 1)
    return u, v
```

With a **static** `num_iters`, the time it takes to compile gets exponentially longer with `num_iters`. 

Minimal reproducer:
```python
import time
import torch

@torch.compile
def sinkhorn_2(Mr, u, v):
    for _ in range(2):
        v = torch.logsumexp(Mr + u.unsqueeze(1), 0)
        u = torch.logsumexp(Mr + v.unsqueeze(0), 1)
    return u, v

@torch.compile
def sinkhorn_3(Mr, u, v):
    for _ in range(3):
        v = torch.logsumexp(Mr + u.unsqueeze(1), 0)
        u = torch.logsumexp(Mr + v.unsqueeze(0), 1)
    return u, v

@torch.compile
def sinkhorn_4(Mr, u, v):
    for _ in range(4):
        v = torch.logsumexp(Mr + u.unsqueeze(1), 0)
        u = torch.logsumexp(Mr + v.unsqueeze(0), 1)
    return u, v

@torch.compile
def sinkhorn_5(Mr, u, v):
    for _ in range(5):
        v = torch.logsumexp(Mr + u.unsqueeze(1), 0)
        u = torch.logsumexp(Mr + v.unsqueeze(0), 1)
    return u, v

@torch.compile
def sinkhorn_6(Mr, u, v):
    for _ in range(6):
        v = torch.logsumexp(Mr + u.unsqueeze(1), 0)
        u = torch.logsumexp(Mr + v.unsqueeze(0), 1)
    return u, v

M = torch.randn(256, 256, device='cuda', requires_grad=True)
u = torch.zeros(256, device=M.device, dtype=M.dtype)
v = torch.zeros(256, device=M.device, dtype=M.dtype)

sinkhorn_fn = [sinkhorn_2, sinkhorn_3, sinkhorn_4, sinkhorn_5, sinkhorn_6]

for i, f in enumerate(sinkhorn_fn):
    start_time = time.time()
    f(M, u, v)
    torch.cuda.synchronize()
    print(f'Iteration {i} took {time.time() - start_time:.4f} seconds')
```

Output:
```
Iteration 0 took 1.1647 seconds                                                                                                                                                                                    
Iteration 1 took 0.3771 seconds                                                                                                                                                                                    
Iteration 2 took 5.8516 seconds                                                                                                                                                                                    
Iteration 3 took 141.5683 seconds
```
The one with 6 iterations did not finish within 10 minutes.

This issue **does not occur** if I do any one of the following:
1. Set `requires_grad` of `M` to False
2. Disable torch.compile
3. Use `torch.sum` rather than `torch.logsumexp`
4. Bring `+u` and `+v` out of `logsumexp`, i.e., Replace `torch.logsumexp(Mr + u.unsqueeze(1), 0)` with `torch.logsumexp(Mr, 0) + u.unsqueeze(1)`
5. Remove either one of the updates for `u` or `v`
6. Remove `Mr`

### Error logs

[tlparse.zip](https://github.com/user-attachments/files/18699450/tlparse.zip) is generated for the code below (5 iterations)

```python
import torch

@torch.compile
def sinkhorn(Mr, u, v):
    for _ in range(5):
        v = torch.logsumexp(Mr + u, 0)
        u = torch.logsumexp(Mr + v, 1)
    return u, v

M = torch.randn(256, 256, device='cuda', requires_grad=True)
u = torch.zeros(256, device=M.device, dtype=M.dtype)
v = torch.zeros(256, device=M.device, dtype=M.dtype)

sinkhorn(M, u, v)
```

### Versions

Collecting environment information...
PyTorch version: 2.6.0+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.12.8 | packaged by conda-forge | (main, Dec  5 2024, 14:24:40) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-126-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA L40S
GPU 1: NVIDIA L40S
GPU 2: NVIDIA L40S
GPU 3: NVIDIA L40S
GPU 4: NVIDIA L40S
GPU 5: NVIDIA L40S
GPU 6: NVIDIA L40S
GPU 7: NVIDIA L40S

Nvidia driver version: 550.90.12
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        52 bits physical, 57 bits virtual
Byte Order:                           Little Endian
CPU(s):                               256
On-line CPU(s) list:                  0-255
Vendor ID:                            AuthenticAMD
Model name:                           AMD EPYC 9554 64-Core Processor
CPU family:                           25
Model:                                17
Thread(s) per core:                   2
Core(s) per socket:                   64
Socket(s):                            2
Stepping:                             1
Frequency boost:                      enabled
CPU max MHz:                          3762.9880
CPU min MHz:                          1500.0000
BogoMIPS:                             6199.67
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid overflow_recov succor smca fsrm flush_l1d
Virtualization:                       AMD-V
L1d cache:                            4 MiB (128 instances)
L1i cache:                            4 MiB (128 instances)
L2 cache:                             128 MiB (128 instances)
L3 cache:                             512 MiB (16 instances)
NUMA node(s):                         2
NUMA node0 CPU(s):                    0-63,128-191
NUMA node1 CPU(s):                    64-127,192-255
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Mitigation; safe RET
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-cusparselt-cu12==0.6.2
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] open_clip_torch==2.29.0
[pip3] optree==0.14.0
[pip3] pytorch-lightning==1.9.5
[pip3] pytorchvideo==0.1.5
[pip3] torch==2.6.0
[pip3] torch-stoi==0.2.3
[pip3] torchaudio==2.6.0
[pip3] torchcde==0.2.5
[pip3] torchcfm==1.0.0
[pip3] torchdiffeq==0.2.5
[pip3] torchdyn==1.0.6
[pip3] torchlibrosa==0.1.0
[pip3] torchmetrics==1.6.1
[pip3] torchsde==0.2.6
[pip3] torchvision==0.21.0
[pip3] triton==3.2.0
[conda] libopenvino-pytorch-frontend 2024.4.0             h5888daf_2    conda-forge
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] nvidia-cublas-cu12        12.4.5.8                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.4.127                 pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.4.127                 pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.4.127                 pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.1.0.70                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.2.1.3                 pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.5.147               pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.6.1.9                 pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.3.1.170               pypi_0    pypi
[conda] nvidia-cusparselt-cu12    0.6.2                    pypi_0    pypi
[conda] nvidia-nccl-cu12          2.21.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.4.127                 pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.4.127                 pypi_0    pypi
[conda] open-clip-torch           2.29.0                   pypi_0    pypi
[conda] optree                    0.14.0                   pypi_0    pypi
[conda] pytorch-lightning         1.9.5                    pypi_0    pypi
[conda] pytorchvideo              0.1.5                    pypi_0    pypi
[conda] torch                     2.6.0                    pypi_0    pypi
[conda] torch-stoi                0.2.3                    pypi_0    pypi
[conda] torchaudio                2.6.0                    pypi_0    pypi
[conda] torchcde                  0.2.5                    pypi_0    pypi
[conda] torchcfm                  1.0.0                    pypi_0    pypi
[conda] torchdiffeq               0.2.5                    pypi_0    pypi
[conda] torchdyn                  1.0.6                    pypi_0    pypi
[conda] torchlibrosa              0.1.0                    pypi_0    pypi
[conda] torchmetrics              1.6.1                    pypi_0    pypi
[conda] torchsde                  0.2.6                    pypi_0    pypi
[conda] torchvision               0.21.0                   pypi_0    pypi
[conda] triton                    3.2.0                    pypi_0    pypi

cc @chauhang @penguinwu @oulgen @jamesjwu @aorenste @anijain2305 @laithsakka

**Comment by williamwen42** [comment]
Can you repro on the nightlies? I cannot repro the exponential compile time locally.

**Comment by hkchengrex** [comment]
Ah, this issue does not occur on nightly. Thanks for the pointer! I guess I will wait for the next release :)

**Comment by hkchengrex** [comment]
For reference, compiling it for 1000 iterations (which is what I need) takes around 650 seconds on nightly. This is faster than the 6-iteration version on 2.6.0.
